{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goal: To create a phylogenetic guide tree for a set of sequences\n",
    "\n",
    "## General Code flow:\n",
    "\n",
    "The program execution starts from the main function defined at the end. <br />\n",
    "There is also a function profile() which profiles the performance of the application for a set of sequences.<br />\n",
    "\n",
    "The code can logically be divided into 3 sections:<br />\n",
    "I> Get a list of Local Multiple Alignments (Match list).<br />\n",
    "II> Calculate Edit Distance Matrix.<br />\n",
    "III> Create and output the phylogenetic guide tree.<br />\n",
    "\n",
    "Each part has a markdown giving a brief overview of the algorithm.<br />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I. The following part implements construction of pairwise match list based on the seed and extend algorithm\n",
    "\n",
    "\n",
    "### Goal:  For a given set of sequences generate a list of matches (Local Multiple Alignments) that would serve as potential anchors.\n",
    "\n",
    "### Algorithm:\n",
    "##### GET_PAIRWISE_MATCHES()\n",
    "1>  For each sequence: <br />\n",
    "2> \t&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**GET_MER_MAP_LIST()** and store them in a hash with open chaining.<br />\n",
    "3>  For each key in the hash<br />\n",
    "4> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove the element if it does not have any matches.<br />\n",
    "4>  For the remaining elements in the hash<br />\n",
    "5> \t&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\t**EXTEND_MATCHES()**<br />\n",
    "\n",
    "##### GET_MER_MAP_LIST():\n",
    "1> Calculate the *default seed weight* by **log(average length of sequences/1.5)**. <br />\n",
    "2> Get a palindromic seed corresponding to the default seed weight.<br />\n",
    "3> Let l be the length of the palindromic spaced seed with m degenerate positions.<br />\n",
    "4> Let N be the length of the sequence for which mer map list has to be generated.<br />\n",
    "5> We apply the seed mask of length l at each position of the sequence to get N - l + 1 mers.<br />\n",
    "6> At each position we take the lexicographically smaller of the forward and the reverse complement subsequence and store it into a hash map of mers<br />\n",
    "\n",
    "##### EXTEND_MATCHES()\n",
    "1>\tFor the matching mer in each sequence:<br />\n",
    "2>\t&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Extend it in the forward and reverse direction until a mismatch occurs.<br />\n",
    "3>\t&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Update the mer positions in  the map to obtain the final match list.<br />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function stores a palindromic spaced seed family for weights between 1-32.\n",
    "\n",
    "# Goal: Get seed mask based on weight and palindromic seed family\n",
    "# Input: \"weight\" which is the weight of the mer\n",
    "# Output: An array of palindromic seed masks having the given weight\n",
    "function get_seed_mask(weight::Int64)\n",
    "    no_seeds = zeros(UInt32, 0)\n",
    "    seed_mask_3 =  [\n",
    "        0x0,0xb,   \n",
    "        0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_4 =  [\n",
    "        0x0,0x3b,  \n",
    "        0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_5 =  [\n",
    "        0x0,0x6b,  \n",
    "        0x0,0x139,    \n",
    "        0x0,0x193,\n",
    "        0x0,0x6b, \n",
    "        0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_6 =  [\n",
    "        0x0,0x58D, #0b10110001101,\n",
    "        0x0,0x653, #0b11001010011,\n",
    "        0x0,0x1AB, #0b110101011,\n",
    "        0x0,0xdb,  #0b11011011,\n",
    "        0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_7 =  [\n",
    "        0x0,0x1953, #0b1100101010011\n",
    "        0x0,0x588d, #0b101100010001101\n",
    "        0x0,0x688b, #0b110100010001011\n",
    "        0x0,0x17d,  #0b101111101,\n",
    "        0x0,0x164d, #0b1011001001101,\n",
    "        0x0,0x0,  0x0,0x0\n",
    "    ]   \n",
    "    seed_mask_8 =  [\n",
    "        0x0,0x3927, #0b11100100100111,\n",
    "        0x0,0x1CA7, #0b1110010100111,\n",
    "        0x0,0x6553, #0b110010101010011,\n",
    "        0x0,0xb6d,  #0b101101101101,\n",
    "        0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_9 =  [\n",
    "        0x0,0x7497,  #0b111010010010111,\n",
    "        0x0,0x1c927, #0b11100100100100111,\n",
    "        0x0,0x72a7,  #0b111001010100111,\n",
    "        0x0,0x6fb,   #0b11011111011,\n",
    "        0x0,0x16ed,  #0b1011011101101,\n",
    "        0x0,0x0\n",
    "    ]\n",
    "    seed_mask_10 = [\n",
    "        0x0,0x1d297, #0b11101001010010111,\n",
    "        0x0,0x3A497, #0b111010010010010111,\n",
    "        0x0,0xE997,  #0b1110100110010111,\n",
    "        0x0,0x6D5B,  #0b110110101011011,\n",
    "        0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_11 = [\n",
    "        0x0,0x7954f,  #0b11110010101001111,\n",
    "        0x0,0x75257,  #0b1110101001001010111,\n",
    "        0x0,0x1c9527, #0b111001001010100100111,\n",
    "        0x0,0x5bed,   #0b101101111101101,\n",
    "        0x0,0x5b26d,  #0b1011011001001101101,\n",
    "        0x0,0x0\n",
    "    ]\n",
    "    seed_mask_12 = [\n",
    "        0x0,0x7954f,  #0b1111001010101001111,\n",
    "        0x0,0x3D32F,  #0b111101001100101111,\n",
    "        0x0,0x768B7,  #0b1110110100010110111,\n",
    "        0x0,0x5B56D,  #0b1011011010101101101,\n",
    "        0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_13 = [\n",
    "        0x0,0x792a4f, #0b11110010010101001001111,\n",
    "        0x0,0x1d64d7, #0b111010110010011010111,\n",
    "        0x0,0x1d3597, #0b111010011010110010111,\n",
    "        0x0,0x1b7db,  #0b11011011111011011,\n",
    "        0x0,0x75ad7,  #0b1110101101011010111,\n",
    "        0x0,0x0\n",
    "    ]\n",
    "    seed_mask_14 = [\n",
    "        0x0,0x1e6acf, #0b111100110101011001111,\n",
    "        0x0,0xF59AF,  #0b11110101100110101111,\n",
    "        0x0,0x3D4CAF, #0b1111010100110010101111,\n",
    "        0x0,0x35AD6B, #0b1101011010110101101011,\n",
    "        0x0,0x0,  0x0,0x0        \n",
    "    ]\n",
    "    seed_mask_15 = [\n",
    "        0x0,0x7ac9af, #0b11110101100100110101111\n",
    "        0x0,0x7b2a6f, #0b11110110010101001101111\n",
    "        0x0,0x79aacf, #0b11110011010101011001111\n",
    "        0x0,0x16df6d, #0b101101101111101101101\n",
    "        0x0,0x6b5d6b, #0b11010110101110101101011\n",
    "        0x0,0x0\n",
    "    ]\n",
    "    seed_mask_16 = [\n",
    "        0x0,0xf599af, #0b111101011001100110101111,\n",
    "        0x0,0xEE5A77, #0b111011100101101001110111,\n",
    "        0x0,0x7CD59F, #0b11111001101010110011111,\n",
    "        0x0,0xEB5AD7, #0b111010110101101011010111,\n",
    "        0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_17 = [\n",
    "        0x0,0x6dbedb, #0b11011011011111011011011,\n",
    "        0x0,0x0, 0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0       \n",
    "    ]\n",
    "    seed_mask_18 = [\n",
    "        0x0,0x3E6B59F, #0b11111001101011010110011111,\n",
    "        0x0,0x3EB335F, #0b11111010110011001101011111,\n",
    "        0x0,0x7B3566F, #0b111101100110101011001101111,\n",
    "        0x0,0x0,  0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_19 = [\n",
    "        0x0,0x7b974ef, #0b111101110010111010011101111\n",
    "        0x0,0x7d6735f, #0b111110101100111001101011111\n",
    "        0x0,0x1edd74f, #0b1111011011101011101101111\n",
    "        0x0,0x0,  0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_20 = [\n",
    "        0x0,0x1F59B35F,#0b11111010110011011001101011111,\n",
    "        0x0,0x3EDCEDF, #0b11111011011100111011011111,\n",
    "        0x0,0xFAE675F, #0b1111101011100110011101011111,\n",
    "        0x0,0x0,  0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_21 = [\n",
    "        0x0,0x7ddaddf, #0b111110111011010110111011111,\n",
    "        0x0,0xaeb3f,   #0b11111100110101110101100111111,\n",
    "        0x0,0x7eb76bf, #0b111111010110111011010111111,\n",
    "        0x0,0x0,  0x0,0x0,  0x0,0x0\n",
    "    ]\n",
    "    seed_mask_22 = [0x0,0x003fffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_23 = [0x0,0x007fffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_24 = [0x0,0x00ffffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_25 = [0x0,0x01ffffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_26 = [0x0,0x03ffffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_27 = [0x0,0x07ffffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_28 = [0x0,0x0fffffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_29 = [0x0,0x1fffffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_30 = [0x0,0x3fffffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    seed_mask_31 = [0x0,0x7fffffff,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0,  0x0,0x0]\n",
    "    \n",
    "    seed_mask = [\n",
    "        no_seeds,\n",
    "        no_seeds,\n",
    "        no_seeds,\n",
    "        seed_mask_3,\n",
    "        seed_mask_4,\n",
    "        seed_mask_5,\n",
    "        seed_mask_6,\n",
    "        seed_mask_7,\n",
    "        seed_mask_8,\n",
    "        seed_mask_9,\n",
    "        seed_mask_10,\n",
    "        seed_mask_11,\n",
    "        seed_mask_12,\n",
    "        seed_mask_13,\n",
    "        seed_mask_14,\n",
    "        seed_mask_15,\n",
    "        seed_mask_16,\n",
    "        seed_mask_17,\n",
    "        seed_mask_18,\n",
    "        seed_mask_19,\n",
    "        seed_mask_20,\n",
    "        seed_mask_21,\n",
    "        seed_mask_22,\n",
    "        seed_mask_23,\n",
    "        seed_mask_24,\n",
    "        seed_mask_25,\n",
    "        seed_mask_26,\n",
    "        seed_mask_27,\n",
    "        seed_mask_28,\n",
    "        seed_mask_29,\n",
    "        seed_mask_30,\n",
    "        seed_mask_31,\n",
    "    ]\n",
    "    return seed_mask[weight]\n",
    "end;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input: Weight of the seed\n",
    "# Output: A seed mask having all ones.\n",
    "function get_solid_seed(weight::Int64)\n",
    "    seed = 1\n",
    "    seed <<= weight\n",
    "    return seed\n",
    "end;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_seed (generic function with 1 method)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Input: Weight of the seed and default seed rank\n",
    "# Output: Seed with given weight\n",
    "function get_seed(weight::Int64, seed_rank::Int64)\n",
    "    mask = get_seed_mask(weight + 1)\n",
    "    low = mask[seed_rank*2 + 2]\n",
    "    if (low == 0)\n",
    "        return get_solid_seed(weight)\n",
    "    end\n",
    "    high = mask[seed_rank*2 + 1]\n",
    "    seed = 0\n",
    "    seed |= high\n",
    "    seed <<=32\n",
    "    seed |= low\n",
    "    return seed\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_default_mer_size (generic function with 1 method)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# From the paper and code, mer_size is calculated by lg_2(totalLen/num_seq)/1.5\n",
    "# If mer_size is even, increment by 1\n",
    "# mer_size is bounded by [5, 31]\n",
    "function get_default_mer_size(total_len::Int64, num_seq::Int64)\n",
    "    mer_size = round(Int64, (log2(round(Int64, total_len/num_seq)))/1.5)\n",
    "    if iseven(mer_size)\n",
    "        mer_size += 1\n",
    "    end\n",
    "    if mer_size < 5\n",
    "        mer_size = 5\n",
    "    end\n",
    "    if mer_size > 31\n",
    "        mer_size = 31\n",
    "    end\n",
    "    return mer_size\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "function get_default_seed(sequences, mer_size, coding_seed)\n",
    "    totalLen = sum(length.(sequences))\n",
    "    numSeq = length(sequences)\n",
    "    if mer_size == 0\n",
    "        mer_size = get_default_mer_size(totalLen, numSeq)\n",
    "    end\n",
    "    default_seed = get_seed(mer_size, coding_seed)\n",
    "end;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_seed_length (generic function with 1 method)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function get_seed_length(seed::Int64)\n",
    "    leftIndex = 64 - leading_zeros(seed)\n",
    "    rightIndex = trailing_zeros(seed)\n",
    "    seedLength = leftIndex - rightIndex\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "apply_seed_mask (generic function with 1 method)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function apply_seed_mask(str::String, default_seed::Int64)\n",
    "    bin_seed_mask = digits(default_seed, base=2)\n",
    "    mer = \"\"\n",
    "    for i in 1:get_seed_length(default_seed)\n",
    "        if bin_seed_mask[i] == 1\n",
    "            mer = mer*str[i]\n",
    "        end\n",
    "    end\n",
    "    return mer\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "rev_map = Dict('A' => 'T', 'T' => 'A', 'C' => 'G', 'G' => 'C', 'a' => 't', 't' => 'a', 'c' => 'g', 'g' => 'c');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "getFwdOrRevMer (generic function with 1 method)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Input: Mer\n",
    "# Output: Lexicographically smaller among the forward and reverse mers\n",
    "function getFwdOrRevMer(fwd_mer::String)\n",
    "    rev_mer = String(\"\")\n",
    "    for nucleotide in fwd_mer\n",
    "        rev_mer = rev_mer*rev_map[nucleotide]\n",
    "    end\n",
    "    if isless(fwd_mer, rev_mer)\n",
    "        return [fwd_mer, 0]\n",
    "    end\n",
    "    return [rev_mer, 1]\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_sorted_mer_map (generic function with 1 method)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Returns sorted and unique mer_list for a sequence\n",
    "function get_sorted_mer_map(mer_list)\n",
    "    # Sort the mer_list lexicographically\n",
    "    sort!(mer_list, by=x->x[1]);\n",
    "    mer_map = Dict{String, Array}()\n",
    "    for val in mer_list\n",
    "        lc_val = lowercase(val[1])  \n",
    "        if haskey(mer_map, lc_val)\n",
    "            push!(mer_map[lc_val], val[2:end])\n",
    "        else\n",
    "            push!(mer_map, lc_val=>[val[2:end]])\n",
    "        end            \n",
    "    end\n",
    "    # Remove repeats\n",
    "    for val in mer_map\n",
    "        if (length(mer_map[val[1]])) >= 2\n",
    "            delete!(mer_map, val[1])\n",
    "        end\n",
    "    end\n",
    "    return mer_map\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "create_match_list (generic function with 1 method)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Returns a hashmap from \"mer\" => list of [seq_index, mer_direction, mer_start, mer_end]].\n",
    "# eg. of an entry: \"atgtc\" => Array{Any,1}[[4, 0, 1, 7] \n",
    "# Where [4, 0, 1, 7] means it's the mer of the 4th sequence, \n",
    "# having direction 0 or forward strand, starts at positon 1 and ends at positon 7\n",
    "# Hashmap is used to keep track of number of matches.\n",
    "\n",
    "function create_match_list(mer_map_list)\n",
    "    match_list_map = Dict{String, Array}()\n",
    "    for seq_index in 1:length(mer_map_list)\n",
    "        mer_map = mer_map_list[seq_index]\n",
    "        for key_val in mer_map\n",
    "            key = key_val[1]\n",
    "            val = key_val[2][1]\n",
    "            if haskey(match_list_map, key)\n",
    "                push!(match_list_map[key], vcat(seq_index, val))\n",
    "            else\n",
    "                push!(match_list_map, key=>[vcat(seq_index, val)])\n",
    "            end\n",
    "        end\n",
    "    end \n",
    "    \n",
    "    # Remove entries that have no matches\n",
    "    for key_val in match_list_map\n",
    "        if length(key_val[2]) < 2\n",
    "            delete!(match_list_map, key_val[1])\n",
    "        end\n",
    "    end\n",
    "    return match_list_map\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_nucleotide (generic function with 1 method)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Returns forward or reverse complement nucleotide depending on the direction\n",
    "function get_nucleotide(direction::Int64, nucleotide)\n",
    "    if direction == 0\n",
    "        return nucleotide\n",
    "    else\n",
    "        return rev_map[nucleotide]\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "extend_matches (generic function with 1 method)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Goal: Extend matches in the forward and the reverse direction and update the match_list_map with new start and \n",
    "# end positions of the matches\n",
    "\n",
    "# Input: \n",
    "# 1> A set of sequences.\n",
    "# 2> A map from \"mers\" => list of [Sequence_number, Direction, Start_Position, End_Position]\n",
    "\n",
    "function extend_matches(sequences, match_list_map)\n",
    "    FWD_MASK = 0x1\n",
    "    BWD_MASK = 0x2\n",
    "    for value in collect(values(match_list_map))\n",
    "#         println(value)\n",
    "        newValue = []\n",
    "        fwd_step = 0\n",
    "        bwd_step = 0 \n",
    "        ext_dir = FWD_MASK | BWD_MASK\n",
    "        new_start = \"\"\n",
    "        new_end = \"\"\n",
    "        \n",
    "        while (ext_dir != 0)\n",
    "            for mer_ind in 1:length(value)\n",
    "                mer = value[mer_ind]\n",
    "                seq_ind = mer[1]\n",
    "                direction = mer[2]\n",
    "                start_ind = mer[3]\n",
    "                end_ind = mer[4]            \n",
    "                cur_seq = sequences[seq_ind]\n",
    "            \n",
    "                # INITIALIZE\n",
    "                if (mer_ind == 1)\n",
    "                    # Check and increment in the backward direction\n",
    "                    if (ext_dir & BWD_MASK == BWD_MASK && (start_ind - bwd_step > 1))\n",
    "                        bwd_step += 1\n",
    "                    else\n",
    "                        ext_dir &= FWD_MASK\n",
    "                    end\n",
    "                    # Check and increment in the forward direction\n",
    "                    if (ext_dir & FWD_MASK == FWD_MASK && (end_ind + fwd_step < length(cur_seq)))\n",
    "                        fwd_step += 1\n",
    "                    else\n",
    "                        ext_dir &= BWD_MASK\n",
    "                    end\n",
    "                    # Assign new_start and new_end depending on the direction\n",
    "                    new_start = get_nucleotide(direction, cur_seq[start_ind - bwd_step])\n",
    "                    new_end = get_nucleotide(direction, cur_seq[end_ind + fwd_step])\n",
    "                    \n",
    "                    # After initialization move on to the next iteration\n",
    "                    continue\n",
    "                end\n",
    "            \n",
    "                # BACK EXTEND\n",
    "                if (ext_dir & BWD_MASK == BWD_MASK)\n",
    "                    if !(start_ind - bwd_step >= 1 && \n",
    "                            get_nucleotide(direction, cur_seq[start_ind - bwd_step]) == new_start)\n",
    "                        bwd_step -= 1\n",
    "                        new_start = get_nucleotide(direction, cur_seq[start_ind - bwd_step])\n",
    "                        ext_dir &= FWD_MASK\n",
    "                    end\n",
    "                end\n",
    "\n",
    "                # FWD EXTEND\n",
    "                if (ext_dir & FWD_MASK == FWD_MASK)\n",
    "                    if !(end_ind + fwd_step <= length(cur_seq) && \n",
    "                            get_nucleotide(direction, cur_seq[end_ind + fwd_step]) == new_end)\n",
    "                        fwd_step -= 1\n",
    "                        new_end = get_nucleotide(direction, cur_seq[start_ind + fwd_step])\n",
    "                        ext_dir &= BWD_MASK\n",
    "                    end  \n",
    "                end\n",
    "            \n",
    "                # If we are not extending in either direction, break\n",
    "                if (ext_dir == 0)\n",
    "                    # println(\"Not extending in either direction\")\n",
    "                    break\n",
    "                end \n",
    "            end\n",
    "        end\n",
    "        # Make the changes\n",
    "        if fwd_step != 0 || bwd_step != 0\n",
    "#             println(\"fwd_step is: \", fwd_step)\n",
    "#             println(\"bwd_step is: \", bwd_step)\n",
    "            for mer_ind in 1:length(value)\n",
    "                value[mer_ind][3] = value[mer_ind][3] - bwd_step\n",
    "                value[mer_ind][4] = value[mer_ind][4] + fwd_step\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_mer_map_list (generic function with 1 method)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Input: List of sequences and default seed (which is 3 by default)\n",
    "# Output: \n",
    "# Returns a list of hashmaps (\"Mer\" => [Direction, Start-Position, End-Position])\n",
    "# Direction is 0 if the mer is on forward strand and 1 if the mer is on the reverse complement strand\n",
    "# Each entry is a hashmap of mers in each sequence\n",
    "# So length of the returned value = Number of sequences\n",
    "\n",
    "function get_mer_map_list(sequences::Array{String, 1}, default_seed)\n",
    "    mer_map_list = Array{Any,1}()\n",
    "    for seq in sequences\n",
    "        mer_list = []\n",
    "        mer_map = Dict{String, Array}()\n",
    "        len = length(seq) - get_seed_length(default_seed) + 1\n",
    "        for pos in 1:len\n",
    "            str = seq[pos:pos+get_seed_length(default_seed)-1]\n",
    "            fwd_mer = apply_seed_mask(str, default_seed) \n",
    "            mer = vcat(getFwdOrRevMer(fwd_mer), [pos, pos + get_seed_length(default_seed) - 1])\n",
    "            push!(mer_list, mer)\n",
    "        end  \n",
    "        mer_map = get_sorted_mer_map(mer_list)\n",
    "        push!(mer_map_list, mer_map)\n",
    "    end  \n",
    "    return mer_map_list\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_pairwise_matches (generic function with 1 method)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Goal: Given a set of sequences generate a list of maximally unique matches.\n",
    "# Input: A set of strings representing individual sequences.\n",
    "# Output: A list of matches ocurring in two or more sequences. \n",
    "\n",
    "function get_pairwise_matches(sequences::Array{String, 1})\n",
    "    DEFAULT_SEED_RANK = 3\n",
    "    mer_size = 0\n",
    "    default_seed = get_default_seed(sequences, mer_size, DEFAULT_SEED_RANK)\n",
    "\n",
    "    # Apply default_seed to each position of each sequence to get a list of mers for each sequence\n",
    "    mer_map_list = get_mer_map_list(sequences, default_seed)\n",
    "\n",
    "    # match_list_map is a hashmap from \"mer\" => list of [seq_index, mer_direction, mer_start, mer_end]].\n",
    "    # eg. of an entry: \"atgtc\" => Array{Any,1}[[4, 0, 1, 7] \n",
    "    # Where [4, 0, 1, 7] means it's the mer of the 4th sequence, \n",
    "    # having direction 0 or forward strand, starts at positon 1 and ends at positon 7\n",
    "    match_list_map = Dict{String,Array}()\n",
    "    match_list_map = create_match_list(mer_map_list)\n",
    "\n",
    "    # Now, extend matches\n",
    "    extend_matches(sequences, match_list_map)\n",
    "\n",
    "    extended_match_list = Array{Array{Union{Nothing, Array{Int64,1}},1},1}()\n",
    "    for value in collect(values(match_list_map))\n",
    "        row_match_list = Array{Union{Nothing, Array{Int64,1}},1}()\n",
    "        for match_index in 1:length(sequences)\n",
    "            has_match_index = false\n",
    "            for mer in value\n",
    "                if match_index == mer[1]\n",
    "                    push!(row_match_list, mer[2:end])\n",
    "                    has_match_index = true\n",
    "                end\n",
    "            end\n",
    "            if !has_match_index\n",
    "                push!(row_match_list, nothing)\n",
    "            end\n",
    "        end\n",
    "        push!(extended_match_list, row_match_list)\n",
    "    end\n",
    "    return extended_match_list\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## II. The following part implements creation of an edit distance matrix from the pairwise match list of mers\n",
    "\n",
    "### Algorithm: Edit-Distance Algorithm\n",
    "\n",
    "Goal: Find the dissimilarity between pairs of sequences using information from the pairwise matches.\n",
    "\n",
    "Input: A list of the position of the matches in the sequences and a list of the sequences.\n",
    "\n",
    "Output: A n x n matrix (where n is the number of sequences input) of values between 0 and 1. Each value represents the percent of the sequence pair that must be altered to become the identity sequence.\n",
    "\n",
    "Algorithm Flow:\n",
    "\n",
    "Step 1) Sort the matches by descending length.\n",
    "\n",
    "Step 2) Subset the matches array to only contain matches that exist for the current pair of sequences being compared.\n",
    "\n",
    "Step 3) Trim overlapping matches in each sequence.\n",
    "\n",
    "Step 4) Calculate the coverage: Coverage within a match are counted and can be weighted based on the type of match/substitiution; Nucleotides not in a match are counted as dissimilar.\n",
    "\n",
    "Step 5) Set the edit distance for the pairwise genomes.\n",
    "\n",
    "Step 6) Repeat until all pairs have a distance calculated for them and the matrix is filled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "findDistances (generic function with 1 method)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sorts matches to be longest to shortest\n",
    "# Input: Two matches from the current pair of sequences being investigated\n",
    "# Output: The two matches sorted by length\n",
    "function sortMatches(matchA::Array{Union{Nothing, Array{Int64,1}},1}, \n",
    "        matchB::Array{Union{Nothing, Array{Int64,1}},1}) \n",
    "    firstA = findfirst(i->i!=nothing, matchA)\n",
    "    firstB = findfirst(i->i!=nothing, matchB)\n",
    "    lenA = matchA[firstA][3] - matchA[firstA][2]\n",
    "    lenB = matchB[firstB][3] - matchB[firstB][2]\n",
    "    return(lenA < lenB)\n",
    "end\n",
    "\n",
    "# Creating a new array of matches that exist in the current pairwise sequences\n",
    "# Input: The indeces of the current pair of sequences (i and j) and the array of matches\n",
    "# Output: An array of matches for JUST the two sequences currently being compared\n",
    "function keepMatches(i::Int64, j::Int64, matches::Array{Array{Union{Nothing, Array{Int64,1}},1},1}) \n",
    "    cur_matches = Array{Union{Nothing, Array{Int64,1}},1}[]\n",
    "    for match in matches\n",
    "        if match[i]!=nothing && match[j]!=nothing\n",
    "            push!(cur_matches, [deepcopy(match[i]), deepcopy(match[j])])\n",
    "        end\n",
    "    end\n",
    "    return(cur_matches)\n",
    "end\n",
    "\n",
    "# Ensure that overlapping matches are not counted twice in either sequence\n",
    "# Input: Two matches from the array containing only matches of the current sequences being compared\n",
    "# Output: Doesn't return anything--just updates the current matches array\n",
    "function trimMatches(a::Array{Union{Nothing, Array{Int64,1}},1}, \n",
    "        b::Array{Union{Nothing, Array{Int64,1}},1}) \n",
    "\n",
    "    # Checking if the second match overlapped with the first match in either sequence\n",
    "    while ( b[1][2] <= a[1][3] && b[1][2] >= a[1][2] ) || ( b[2][2] <= a[1][3] && b[2][2] >= a[1][3] )\n",
    "        # move the start position up until it no longer overlaps with the other match\n",
    "        b[1][2] += 1\n",
    "        b[2][2] += 1\n",
    "    end\n",
    "    \n",
    "    while ( b[1][3] <= a[1][3] && b[1][3] >= a[1][2] ) || ( b[2][3] <= a[1][3] && b[2][3] >= a[1][3] )\n",
    "        # move the end position down until it no longer overlaps with the other match\n",
    "        b[1][3] -= 1\n",
    "        b[2][3] -= 1\n",
    "    end\n",
    "    \n",
    "    if b[1][3] - b[1][2] <= 0 # a match has been completely removed\n",
    "        b[1] = nothing\n",
    "        b[2] = nothing\n",
    "    end\n",
    "end\n",
    "\n",
    "\n",
    "# Calculates coverage -- can be weighted on which nucleotide has been substituted\n",
    "# Input: The two strains being compared and the matches in those two srains\n",
    "# Output: The number of nucleotides contained in a match which are identical in both strains\n",
    "#         This can be tweaked to weight matches and mismatches in a more biologically-oriented way\n",
    "function ntWeight(g1::String, g2::String, matches::Array{Array{Union{Nothing, Array{Int64,1}},1},1})\n",
    "    coverageA = 0\n",
    "    coverageB = 0\n",
    "    \n",
    "    strong = 1\n",
    "    weak = .95\n",
    "    purine_transition = .5\n",
    "    pyrimidine_transition = .5\n",
    "    transversion = 0\n",
    "    \n",
    "    for m in 1:length(matches)\n",
    "        if matches[m][1] != nothing\n",
    "            ntSeq2 = matches[m][2][2] # starting point nucleotide in the second sequence\n",
    "            for ntSeq1 in matches[m][1][2]:matches[m][1][3] # starting point nucleotide in the first sequence\n",
    "                if (g1[ntSeq1]=='a' && g2[ntSeq2]=='a') || (g1[ntSeq1]=='t' && g2[ntSeq2]=='t') # strong\n",
    "                    coverageA += strong\n",
    "                    coverageB += strong\n",
    "                elseif (g1[ntSeq1]=='g' && g2[ntSeq2]=='g') || (g1[ntSeq1]=='c' && g2[ntSeq2]==\"c\") # weak\n",
    "                    coverageA += weak\n",
    "                    coverageB += weak\n",
    "                elseif (g1[ntSeq1]=='a' && g2[ntSeq2]=='g') || (g1[ntSeq1]=='g' && g2[ntSeq2]=='a') # purine\n",
    "                    coverageA += purine_transition\n",
    "                    coverageB += purine_transition\n",
    "                elseif (g1[ntSeq1]=='c' && g2[ntSeq2]=='t') || (g1[ntSeq1]=='t' && g2[ntSeq2]=='c') #pyrimidine\n",
    "                    coverageA += pyrimidine_transition\n",
    "                    coverageB += pyrimidine_transition\n",
    "                else # transversion substitution\n",
    "                    coverageA += transversion\n",
    "                    coverageB += transversion\n",
    "                end                \n",
    "                ntSeq2 += 1                \n",
    "            end\n",
    "        end\n",
    "    end    \n",
    "    return coverageA, coverageB\n",
    "    \n",
    "end\n",
    "\n",
    "\n",
    "########## Begin Processing Matches to Find Edit Distances ##########\n",
    "\n",
    "function findDistances(sequences, matches)    \n",
    "    # Make sure that the incoming sequences are all lowercase\n",
    "    seqnum=1\n",
    "    for seq in sequences\n",
    "        sequences[seqnum] = lowercase(seq)\n",
    "        seqnum += 1\n",
    "    end    \n",
    "    \n",
    "    #initialize output matrix\n",
    "    edit_distance_matrix = zeros(Float64, (length(sequences), length(sequences))) \n",
    "\n",
    "    # Step 1) Sort the matches by descending length \n",
    "    sort!(matches, lt=sortMatches, rev=true)\n",
    "\n",
    "    # Pairwise Comparison\n",
    "    for i in 1:length(sequences)                \n",
    "        for j in (i+1):length(sequences)\n",
    "            # Step 2) Subset the matches array to only contain matches that exist \n",
    "            #         for the current pair of sequences being compared\n",
    "            cur_matches = keepMatches(i,j,matches)\n",
    "            \n",
    "            for k in 1:length(cur_matches)\n",
    "                # for the case that an entire match was trimmed away in a previous round\n",
    "                if cur_matches[k][1] == nothing \n",
    "                    continue\n",
    "                end\n",
    "                for l in (k+1):length(cur_matches)\n",
    "                    if cur_matches[l][1] == nothing # for the case that an entire match was trimmed away\n",
    "                        continue\n",
    "                    end\n",
    "                    # Step 3) Trim overlapping matches in each sequence\n",
    "                    trimMatches(cur_matches[k], cur_matches[l])\n",
    "                end\n",
    "            end\n",
    "\n",
    "            # Step 4) Calculate the coverage:\n",
    "            #         Coverage within a match can be weighted based on the type of match/substitiution\n",
    "            #         Nucleotides not in a match are counted as dissimilar.\n",
    "            coverage = ntWeight(sequences[i], sequences[j], cur_matches)\n",
    "            coverageA = coverage[1]\n",
    "            coverageB = coverage[2]\n",
    "\n",
    "            percentA = coverageA / length(sequences[i])\n",
    "            percentB = coverageB / length(sequences[j])\n",
    "            percentAB = (percentA + percentB) / 2   # average the coverage\n",
    "            editPercent = 1 - percentAB \n",
    "            \n",
    "            # Step 5) Set the edit distance for the pairwise genomes\n",
    "            edit_distance_matrix[i, j] = editPercent\n",
    "\n",
    "            # Step 6) Repeat until all pairs have a distance calculated for them and the matrix is filled\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # Convert to a matrix from array of arrays and fill the lower half\n",
    "    row_len = size(edit_distance_matrix, 1)\n",
    "    for i = 1:row_len, j=1:row_len\n",
    "        if i>j\n",
    "            edit_distance_matrix[i, j] = edit_distance_matrix[j, i]\n",
    "        end\n",
    "    end\n",
    "    return edit_distance_matrix\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## III. The following part pertains to construction of the  phylogenetic guide tree from the edit distance matrix\n",
    "\n",
    "### Algorithm:\n",
    "\n",
    "## TODO: \n",
    "1> Add more comments. <br /> \n",
    "2> Make function names and variables user friendly example Q, qmat, q() etc. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: replacing module TreeMod.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "write_file (generic function with 2 methods)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Defining a \"Tree\" composite type\n",
    "module TreeMod\n",
    "    mutable struct Tree\n",
    "        name::String\n",
    "        newick::String\n",
    "        par::Any\n",
    "        branch::Any\n",
    "        chil::Dict\n",
    "        connections::Dict\n",
    "    \n",
    "        Tree(name, newick) = new(name, newick)\n",
    "        function Base.show(io::IO, t::Tree)\n",
    "            println(t.name)\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "function MakeAnchorTree(D, debug::Bool = false)\n",
    "    # Make Q1 Matrix\n",
    "    function QMat(A)\n",
    "        m = size(A,1)\n",
    "        Q = copy(A)\n",
    "        function q(i,j)\n",
    "            if i==j\n",
    "                return 0\n",
    "            else\n",
    "                n = size(D,1) - 1\n",
    "                dij = D[i,j]\n",
    "                sum_dik = sum(D[i,:])\n",
    "                sum_djk = sum(D[:,j])\n",
    "                return (m-2) * dij - sum_dik - sum_djk\n",
    "            end\n",
    "        end\n",
    "    \n",
    "        for i = 1:m\n",
    "            for j = 1:m\n",
    "                Q[i,j] = q(i,j)\n",
    "            end\n",
    "        end\n",
    "        return Q\n",
    "    end\n",
    "\n",
    "    # Get the distance between a node k and nodes f and g\n",
    "    function dist(k,f,g)\n",
    "        dfk = D[f,k]\n",
    "        dgk = D[g,k]\n",
    "        dfg = D[f,g]\n",
    "        return (1/2) * (dfk + dgk - dfg)\n",
    "    end\n",
    "\n",
    "    # Create the nodes, add to the tree, update their connections\n",
    "    function make_nodes(f,g,index_trees)\n",
    "        # Create the tree nodes for f and g, and then create a new neighbor node u that connects the two\n",
    "        s = index_trees[f]\n",
    "        t = index_trees[g]\n",
    "        uname = \"neighbor\"* string(r)\n",
    "        u = TreeMod.Tree(uname, uname) # name neighbor nodes sequentially\n",
    "\n",
    "        # compute distance from f and g to u -- returns array of dist(f), dist(g)\n",
    "        node_dists = deltas(f,g)\n",
    "        #println(\"node distances: \", node_dists)\n",
    "        \n",
    "        # update connections list with branch lengths\n",
    "        if isdefined(u, :connections)\n",
    "            merge!(u.connections, Dict(s => node_dists[1], t => node_dists[2]))\n",
    "        else\n",
    "            u.connections = Dict(s => node_dists[1], t => node_dists[2])\n",
    "        end\n",
    "\n",
    "        if isdefined(s, :connections)\n",
    "            merge!(s.connections, Dict(u => node_dists[1]))\n",
    "        else\n",
    "            s.connections = Dict(u => node_dists[1])\n",
    "        end\n",
    "        \n",
    "        if isdefined(t, :connections)\n",
    "            merge!(t.connections, Dict(u => node_dists[2]))\n",
    "        else\n",
    "            t.connections = Dict(u => node_dists[2])\n",
    "        end\n",
    "        # update the new node's Newick string\n",
    "        newick = \"($(s.newick):$(node_dists[1]), $(t.newick):$(node_dists[2]))$(u.newick)\"\n",
    "        u.newick = newick\n",
    "        #println(\"new node's newick: \",newick)\n",
    "\n",
    "        # update list of trees and new index list\n",
    "        #println(index)\n",
    "        #println(index_trees)\n",
    "        index_trees = [index_trees[i] for i in index if i != f && i != g]\n",
    "\n",
    "        pushfirst!(index_trees, u)\n",
    "        push!(tree, u)\n",
    "\n",
    "        return index_trees, tree\n",
    "    end\n",
    "\n",
    "    # Calculate distance nodes f and g to the connecting neighbor node\n",
    "    function deltas(f,g)\n",
    "        dfg = D[f,g]\n",
    "        sum_dfk = sum(D[f,:])\n",
    "        sum_dgk = sum(D[:,g])\n",
    "        deltaf = ((1/2) * dfg + 1/(2(m-2)) * (sum_dfk - sum_dgk))\n",
    "        deltag = dfg - deltaf\n",
    "        return [deltaf,deltag]\n",
    "    end\n",
    "\n",
    "    index = [i for i in 1:n]\n",
    "    index_ids = [\"seq\"*string(i) for i in index]\n",
    "\n",
    "    global index_trees = [TreeMod.Tree(i, i) for i in index_ids]\n",
    "    global tree\n",
    "    tree = copy(index_trees)\n",
    "               \n",
    "    # Start out with n nodes that need to be connected and an n+1xn+1 matrix D\n",
    "    m = n\n",
    "    D = D\n",
    "    r = 1\n",
    "\n",
    "    # Until we get down to 2 nodes\n",
    "    while m > 3\n",
    "        if debug\n",
    "            println(\"\\nNEW ITERATION!!!\\n====================\")\n",
    "            println(\"Number of nodes still to connect (m): \", m,\" ...at iteration: \",r)\n",
    "            println(\"\\nFinding sequences to connect\\n====================\")\n",
    "            println(\"Converting absolute distances from D to proportional (?) distances in Q\")\n",
    "            display(D)\n",
    "        end\n",
    "                                \n",
    "        Q = QMat(D)\n",
    "        if debug display(Q) end\n",
    "\n",
    "        # Find the smallest distance between two sequences (smallest Q value)\n",
    "        farthest_seqs = argmin(Q)\n",
    "\n",
    "        # Indices of f and g are the indices of D and Q where these values are\n",
    "        f = farthest_seqs[1]\n",
    "        g = farthest_seqs[2]\n",
    "\n",
    "        if debug println(\"farthest apart nodes are f ($(index_ids[f])) at index $f and g ($(index_ids[g])) at index $g\") end\n",
    "        \n",
    "        # make the nodes for f and g, connect them to a new internal node, and add distances and connections\n",
    "        if debug println(\"\\nConnecting the sequences\\n====================\") end\n",
    "        index_trees, tree = make_nodes(f,g,index_trees)\n",
    "        \n",
    "        # if we have more nodes left to connect, continue on and prepare new distance matrix\n",
    "        if debug println(\"\\nUpdating distance matrix and indices\\n====================\") end\n",
    "        # Initialize a new n-1 x n-1 distance matrix of zeros (because have joined two nodes)\n",
    "        D_new = zeros(m-1,m-1)\n",
    "\n",
    "        index_new = [i for i in index if i != g && i != f]\n",
    "        index_ids_new = [index_ids[i] for i in index if i != g && i != f]\n",
    "\n",
    "        # Make a new array without the rows and columns of nodes we have already connected\n",
    "        D_old = D[setdiff(1:end, (f,g)), setdiff(1:end, (f,g))]\n",
    "\n",
    "        # Distances not affected by the join are the same as in D\n",
    "        # So place the values from the D_old matrix into the D_new matrix, after the first row\n",
    "        # and column, which are reserved for the newly created node.      \n",
    "        for i in 2:m-1\n",
    "            for j in 2:m-1\n",
    "              D_new[i,j] = D_old[i-1, j-1]\n",
    "            end\n",
    "        end\n",
    "\n",
    "        # Calculate distances from the new node to the other nodes\n",
    "        for i in 1:m-2            \n",
    "            old_idx = index_new[i]\n",
    "            thedist = dist(old_idx,f,g)\n",
    "            D_new[1,i+1] = thedist\n",
    "            D_new[i+1,1] = thedist\n",
    "        end\n",
    "\n",
    "        # Update the distance index:\n",
    "        # First, remove the two sequences we have already placed from our index\n",
    "        # and replace with a dummy index called 0 for the joined node\n",
    "        index_clean = [i for i in 1:m-1]\n",
    "        index = index_clean\n",
    "        index_ids = pushfirst!(index_ids_new, \"internal\")\n",
    "\n",
    "        # Make D_new our main distance matrix\n",
    "        D = D_new\n",
    "\n",
    "        # Update m\n",
    "        m -= 1\n",
    "        r += 1\n",
    "    end\n",
    "\n",
    "    # 3 nodes left... join the first two nodes s and t at a new node u\n",
    "    s = index_trees[1]\n",
    "    t = index_trees[2]\n",
    "    uname = \"neighbor\"* string(r)\n",
    "    node_dists = deltas(1,2)\n",
    "    u = TreeMod.Tree(uname, uname) # name neighbor nodes sequentially\n",
    "    \n",
    "    # Add connections between s and u, t and u                                    \n",
    "    if isdefined(s, :connections)\n",
    "        merge!(s.connections, Dict(u => node_dists[1]))\n",
    "    else\n",
    "        s.connections = Dict(u => node_dists[1])\n",
    "    end\n",
    "        \n",
    "    if isdefined(t, :connections)\n",
    "        merge!(t.connections, Dict(u => node_dists[2]))\n",
    "    else\n",
    "        t.connections = Dict(u => node_dists[2])\n",
    "    end\n",
    "\n",
    "    if isdefined(u, :connections)\n",
    "        merge!(u.connections, Dict(s => node_dists[1], t => node_dists[2]))\n",
    "    else\n",
    "        u.connections = Dict(s => node_dists[1], t => node_dists[2])\n",
    "    end\n",
    "\n",
    "    # Then connect the last node r with the newly created node u\n",
    "    r = index_trees[3]\n",
    "    r_u_dist = dist(3, 1, 2)\n",
    "\n",
    "    if isdefined(r, :connections)\n",
    "        merge!(r.connections, Dict(u => r_u_dist))\n",
    "    else\n",
    "        r.connections = Dict(u => r_u_dist)\n",
    "    end\n",
    "\n",
    "    merge!(u.connections, Dict(r => r_u_dist))\n",
    "    \n",
    "    # Add the new node to the tree\n",
    "    push!(tree, u)\n",
    "                                        \n",
    "    return tree\n",
    "end\n",
    "\n",
    "\n",
    "function reconstruct_path(cameFrom, current)\n",
    "    total_path = [current]\n",
    "    total_sum = 0\n",
    "    while current in collect(keys(cameFrom))\n",
    "        current = cameFrom[current]\n",
    "        push!(total_path, current)\n",
    "    end\n",
    "    # println(\"total path: $total_path\")\n",
    "    return total_path\n",
    "\n",
    "end\n",
    "\n",
    "function A_Star(start, goal, tree)\n",
    "    # The set of nodes already evaluated\n",
    "    closedSet = Array{Any, 1}()\n",
    "\n",
    "    # The set of currently discovered nodes that are not evaluated yet.\n",
    "    # Initially, only the start node is known.\n",
    "    openSet = [start]\n",
    "\n",
    "    # For each node, which node it can most efficiently be reached from.\n",
    "    # If a node can be reached from many nodes, cameFrom will eventually contain the\n",
    "    # most efficient previous step.\n",
    "    cameFrom = Dict()\n",
    "\n",
    "    # For each node, the cost of getting from the start node to that node.\n",
    "    # Initialized at infinity\n",
    "    gScore = Dict(i => Inf for i in tree) \n",
    "\n",
    "    # The cost of going from start to start is zero.\n",
    "    gScore[start] = 0\n",
    "\n",
    "    # For each node, the total cost of getting from the start node to the goal\n",
    "    # by passing by that node. That value is partly known, partly heuristic.\n",
    "    fScore = Dict()\n",
    "\n",
    "    # For the first node, that value is completely heuristic.\n",
    "    # But we don't have a heuristic function for this so we set it to 0\n",
    "    fScore[start] = 0\n",
    "\n",
    "    while !isempty(openSet)\n",
    "        # println(\"============\")\n",
    "        \n",
    "        scores = Dict(i => fScore[i] for i in openSet)\n",
    "        current = findmin(scores)[2]\n",
    "        if current == goal\n",
    "            return [reconstruct_path(cameFrom, current), gScore[current]]\n",
    "        end\n",
    "                                                \n",
    "        deleteat!(openSet, findfirst(isequal(current), openSet))\n",
    "        push!(closedSet, current)\n",
    "\n",
    "        # println(\"For neighbors -----\")\n",
    "        for neighbor in keys(current.connections)\n",
    "            # println(neighbor)\n",
    "            if neighbor in closedSet\n",
    "                continue        # Ignore the neighbor which is already evaluated.\n",
    "            end\n",
    "\n",
    "            # The distance from start to a neighbor\n",
    "            tentative_gScore = gScore[current] + current.connections[neighbor]\n",
    "\n",
    "            if !(neighbor in openSet)   # Discover a new node\n",
    "                push!(openSet, neighbor)\n",
    "            elseif tentative_gScore >= gScore[neighbor]\n",
    "                continue     \n",
    "            end\n",
    "\n",
    "            # This path is the best until now. Record it!\n",
    "            cameFrom[neighbor] = current\n",
    "            gScore[neighbor] = tentative_gScore\n",
    "            fScore[neighbor] = gScore[neighbor] + 0\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "function getRootID(midpath)\n",
    "    traveled = 0\n",
    "    for i in 1:length(midpath)-1\n",
    "        current = midpath[i]\n",
    "        next = midpath[i+1]\n",
    "        dist_to_next = current.connections[next]\n",
    "        new_traveled = traveled + dist_to_next\n",
    "        \n",
    "        # If after going to the next node, we won't be at the midpoint yet, do it                              \n",
    "        if new_traveled < midpoint\n",
    "            current = next\n",
    "            traveled = new_traveled\n",
    "        \n",
    "        # If after going to the next node, we will have passed the midpoint, stop and root                                           \n",
    "        elseif new_traveled > midpoint\n",
    "            # Create root node\n",
    "            root = TreeMod.Tree(\"root\", \"root\")\n",
    "            push!(tree, root)\n",
    "                                                    \n",
    "            # Specify the parent => child relationship between root and the two nodes it is. between                                        \n",
    "            root.chil = Dict(current => midpoint-traveled, next=> new_traveled-midpoint)\n",
    "            current.par = Dict(root => midpoint-traveled)\n",
    "            next.par = Dict(root => new_traveled-midpoint)\n",
    "            \n",
    "            # Root node is now between \"curent\" and \"next\" nodes, so remove eachother from connections\n",
    "            delete!(current.connections, next)\n",
    "            delete!(next.connections, current)\n",
    "            \n",
    "            # Return an array with the new root, and its two children\n",
    "            return([current, next, root])\n",
    "            break\n",
    "        \n",
    "        # In the unlikely event that the midpoint is right on the next node, rename and return that as the root\n",
    "        elseif new_traveled == midpoint\n",
    "            next.name = \"root\"\n",
    "            root = current\n",
    "            return(current)\n",
    "            break\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "function update_children(not_an_orphan, childless)\n",
    "    while !isempty(childless)\n",
    "        just_got_a_parent = Array{Any, 1}()\n",
    "        for node in childless\n",
    "            #println(\"parent: $node\")\n",
    "            if isdefined(node, :connections)\n",
    "                for c in keys(node.connections)\n",
    "                    if !(c in not_an_orphan)\n",
    "                        branch_l = node.connections[c]\n",
    "                        c.par = Dict(node => branch_l)\n",
    "                        if isdefined(node, :chil)\n",
    "                            merge!(node.chil, Dict(c => branch_l))\n",
    "                        else\n",
    "                            node.chil = Dict(c => branch_l)\n",
    "                        end\n",
    "                        push!(not_an_orphan, c)\n",
    "                        push!(just_got_a_parent, c)\n",
    "                        push!(childless, c)\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            deleteat!(childless, findfirst(isequal(node), childless))\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "function create_phyogenetic_tree(distanceMatrix, debug = false)\n",
    "    global n                                        \n",
    "    n = size(distanceMatrix, 1)                                  \n",
    "    tree = MakeAnchorTree(distanceMatrix, false)\n",
    "\n",
    "    node_dict = Dict()\n",
    "    for node in tree\n",
    "        merge!(node_dict, Dict(node.name => node))\n",
    "    end\n",
    "\n",
    "    # initiate an array of paths from each leaf\n",
    "    leaves = [t for t in tree if length(keys(t.connections)) == 1]\n",
    "    no_leaves = length(leaves)\n",
    "    leaf_dists = Dict()\n",
    "    leaf_paths = Dict()\n",
    "\n",
    "    for i = 1:no_leaves, j = 1:no_leaves\n",
    "        if i > j\n",
    "            A = A_Star(leaves[i], leaves[j], tree)\n",
    "            leaf_dists[(i,j)] = A[2]\n",
    "            leaf_paths[(i,j)] = A[1]\n",
    "        end\n",
    "    end\n",
    "\n",
    "    max_pairdist = findmax(leaf_dists)\n",
    "    global midpoint                                                    \n",
    "    midpoint = max_pairdist[1]/2\n",
    "\n",
    "    p = max_pairdist[2][1]\n",
    "    q = max_pairdist[2][2]\n",
    "\n",
    "    midpath = leaf_paths[max_pairdist[2]]\n",
    "    rooted = getRootID(midpath)\n",
    "\n",
    "    if length(rooted) > 1\n",
    "        root = rooted[3]\n",
    "        childless = [rooted[1], rooted[2]]\n",
    "    else\n",
    "        root = rooted\n",
    "        childless = Array{Any, 1}(root)\n",
    "    end\n",
    "\n",
    "    # Don't try to parent these connections -- they have already been seeen\n",
    "    not_an_orphan = rooted\n",
    "    update_children(not_an_orphan, childless)\n",
    "\n",
    "    leaves = [i for i in tree if !isdefined(i, :chil)]\n",
    "    for i in tree\n",
    "        if isdefined(i, :par)\n",
    "            p = i.par\n",
    "        else\n",
    "            p = i\n",
    "        end\n",
    "    end\n",
    "\n",
    "    to_n = Array{Any, 1}()\n",
    "    for i in leaves\n",
    "        i.newick = i.name\n",
    "        push!(to_n, collect(keys(i.par))[1])\n",
    "    end\n",
    "\n",
    "    while !isempty(to_n)\n",
    "        p = pop!(to_n)\n",
    "        children = collect(keys(p.chil))\n",
    "        c1 = children[1]\n",
    "        c2 = children[2]                                                                  \n",
    "        p.newick = \"($(c1.newick):$(p.chil[c1]),$(c2.newick):$(p.chil[c2]))\" * (occursin(\"seq\", p.name) ? \"$(p.name)\" : \"\")\n",
    "        if p.name != \"root\"\n",
    "            pushfirst!(to_n, collect(keys(p.par))[1])\n",
    "        end\n",
    "    end\n",
    "    println(root.newick*\";\")\n",
    "    return root.newick*\";\"\n",
    "end\n",
    "\n",
    "function write_file(buffer::String, file_name::String=\"outputfile.nwk\")\n",
    "    output_file = open(file_name, \"w\")\n",
    "    write(output_file, buffer)\n",
    "    close(output_file)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "profile (generic function with 3 methods)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This function profiles the performance of the code for a given set of strings\n",
    "\n",
    "# Output: It generates two files. The first files is in the nwk format which has the resultant phylogentic tree.\n",
    "# The second file contains the time and memory required for generating the tree on the given set of sequences.\n",
    "function profile() \n",
    "    out_file_name = \"profile\"\n",
    "    start_time = time()\n",
    "    mem_alloc = @allocated main()\n",
    "    end_time = time()\n",
    "    elapsed_time = Float16(end_time - start_time)\n",
    "\n",
    "    stats = \"Statistics:\\n\";\n",
    "    stats *= \"ElapsedTime: \"*string(elapsed_time)*\"\\n\"\n",
    "    stats *= \"MemAlloc: \"*string(mem_alloc)*\"\\n\"\n",
    "    write_file(stats, out_file_name*\"_stats.txt\")\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "main (generic function with 1 method)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function main()\n",
    "    sequences = [\n",
    "        \"atcgctcacgacaagcatcagcatcagcatcatacgtgactcaagaaGAATAtcatgttaAATGGTTttactaccccctgtggggatatcgatctactactacgact\",\n",
    "        \"atctagccgatgattctaccacgatctcatactacttagcattagcatcatacgcgatcgggggatatatctctactacgtctcagacacgactccgact\",\n",
    "        \"atcggtcactacttgcatcagcaTTGCTCAATCTGtcagcatgcatacCGTCCACCATAAGAAAAGATGGtcgaggatcatcttattactacccccgcgggggatatatatcttctacgactcagctaAACATTCAAcgattacgact\",\n",
    "        \"atcgttctacgagctcatactactagcatcagcatcagcatCCCATTAgatacgcgatcagcacatcttattactacccccgccggggatatatacagcagcagcatcgactctactacgactt\",\n",
    "        \"atcggtcactactagcatcagatcaccatactcgagcatcatcttattactaccccgcgcatactacgactaccgcggtggatatatatctactaccactcagctacgactacgact\"\n",
    "    ];\n",
    "\n",
    "    # sequences = readlines(open(ARGS[1])) # If sequence file is present as command line input\n",
    "    pairwise_matches = get_pairwise_matches(sequences);\n",
    "    distanceMatrix = findDistances(sequences, pairwise_matches);\n",
    "\n",
    "    phyloTree = String(\"\")\n",
    "    phyloTree = create_phyogenetic_tree(distanceMatrix, false)\n",
    "\n",
    "    write_file(phyloTree)\n",
    "    # write_tree_file(phyloTree, ARGS[1]*output.nwk)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(((seq1:0.30267648171292266,seq4:0.25901895090691446):0.049915093230042784,seq2:0.28715729899184456):0.0016133558442779639,(seq3:0.30966844194122223,seq5:0.24384420532142495):0.04453648884602118);\n"
     ]
    }
   ],
   "source": [
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.0.2",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
